go-controller/pkg/ovn/address_set.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// _update updates the OVN database with the address set's addresses
+func (as *addressSet) setOrClear() {
+	joinedIPs, haveAny := as.joinIPs()
+	if haveAny {
+		_, stderr, err := util.RunOVNNbctl("set", "address_set",
+			as.hashName, fmt.Sprintf("addresses=%s", joinedIPs))
+		if err != nil {
+			klog.Errorf("failed to set address_set, stderr: %q (%v)",
+				stderr, err)
+		}
+	} else {
+		_, stderr, err := util.RunOVNNbctl("clear", "address_set", as.hashName, "addresses")
+		if err != nil {
+			klog.Errorf("failed to clear address_set, stderr: %q (%v)", stderr, err)
+		}
+	}
+}
+
+func (as *addressSet) ReplacePorts(ports map[string]net.IP) {
+	as.Lock()
+	defer as.Unlock()
+	if as.destroyed {
+		return
+	}
+
+	klog.V(5).Infof("ReplacePorts(%s/%s) with %v", as.hashName, as.name, ports)
+	as.ports = ports
+	as.setOrClear()
+}
+
+func (as *addressSet) AddPort(logicalPort string, ip net.IP) error {
+	as.Lock()
+	defer as.Unlock()
+	if as.destroyed || as.ports[logicalPort] != nil {
+		return nil
+	}
+
+	klog.V(5).Infof("AddPort(%s/%s) %s=%s", as.hashName, as.name, logicalPort, ip)
+	as.ports[logicalPort] = ip
+
+	_, stderr, err := util.RunOVNNbctl("add", "address_set", as.hashName, "addresses", ip.String())
+	if err != nil {
+		return fmt.Errorf("failed to add port %q address %q to address_set %q, stderr: %q (%v)",
+			logicalPort, ip, as.hashName, stderr, err)
+	}
+	return nil
+}
+
+func (as *addressSet) DeletePort(logicalPort string) error {
+	as.Lock()
+	defer as.Unlock()
+	if as.destroyed || as.ports[logicalPort] == nil {
+		return nil
+	}
+
+	ip := as.ports[logicalPort]
+	klog.V(5).Infof("DeletePort(%s/%s) %s=%s", as.hashName, as.name, logicalPort, ip)
+	delete(as.ports, logicalPort)
+
+	_, stderr, err := util.RunOVNNbctl("remove", "address_set", as.hashName, "addresses", ip.String())
+	if err != nil {
+		return fmt.Errorf("failed to remove port %q address %q from address_set %q, stderr: %q (%v)",
+			logicalPort, ip, as.hashName, stderr, err)
+	}
+	return nil
+}
+
+func getPodIP(pod *kapi.Pod) net.IP {
+	annotation, _ := util.UnmarshalPodAnnotation(pod.Annotations)
+	if annotation != nil {
+		return annotation.IPs[0].IP
+	}
+	if pod.Status.PodIP != "" {
+		return net.ParseIP(pod.Status.PodIP)
+	}
+	return nil
+}
+
+func (as *addressSet) AddPod(pod *kapi.Pod) error {
+	if ip := getPodIP(pod); ip != nil {
+		return as.AddPort(podLogicalPortName(pod), ip)
+	}
+	return fmt.Errorf("failed to determine pod %s/%s IP", pod.Namespace, pod.Name)
+}
+
+func (as *addressSet) DelPod(pod *kapi.Pod) error {
+	return as.DeletePort(podLogicalPortName(pod))
+}
+
+// PortIterFn will be called once per port and should return an error to stop
+// iteration.
+type PortIterFn func(logicalPort string, ip net.IP)
+
+func (as *addressSet) ForEachPort(iterFn PortIterFn) {
+	as.Lock()
+	defer as.Unlock()
+	if !as.destroyed {
+		for logicalPort, ip := range as.ports {
+			iterFn(logicalPort, ip)
+		}
+	}
+}
+
+func destroyAddressSet(hashName string) {
```

Comment:
re my previous comment about my other branch, I dislike the fact that there are still both `addressSet` methods and independent address-set-related functions in this version